### PR TITLE
Add OpenBLAS 0.3.13

### PR DESCRIPTION
**Pull request by omus**

### CODE REVIEW

**Comment by giordano** [comment]
Still WIP or ready to be merged?

**Comment by omus** [comment]
I believe it should be good. I just copied the 0.3.12 patches for use on 0.3.13 and I wasn't sure if they would apply cleanly.

File: O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch [hunk 1]
```diff
@@ -0,0 +1,17 @@
+--- a/kernel/x86_64/dasum_microk_skylakex-2.c
++++ b/kernel/x86_64/dasum_microk_skylakex-2.c
+@@ -39,10 +39,10 @@
```
Comment (giordano) [inline O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch:3]: Uhm, I believe this patch is wrong (I made it, so I have only to blame myself).  I need to double-check it later

Comment (giordano) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch:3]: Yes, this is a [GCC bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87467).  The patch tried to accommodate the bugged signature of the function, which is wrong.  We should either [patch the header file on-the-fly](https://github.com/gcc-mirror/gcc/commit/23bce99cbe7016a04e14c2163ed3fe6a5a64f4e2#diff-c3bbd1dcd7c953ef6d719969fb187e31209ef3d7851a84448aad949aea2e8106) or require GCC v9, where this should be already fixed.  (I think waiting to fix all of our bugged shards is not really an option.)  I'll ask on Slack

Comment (omus) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch:3]: Should this patch be added to Julia for source builds of OpenBLAS?

Comment (giordano) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch:3]: Not sure how that would work, I think you don't want (and most often can't) apply a patch to GCC header files.

Comment (omus) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch:3]: Ah, I totally missed that these were kernel patches

Comment (giordano) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch:3]: Ok, the poll on Slack suggests to patch the header file `gcc/config/i386/avx512fintrin.h` of GCC.  This patch is definitely wrong and should be removed in any case.  In the future we'll likely backport the patch to all of our versions of GCC affected by the issue.

Comment (omus) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch]: I've removed this patch from OpenBLAS 0.3.13. Let me know if I should also remove it from the other versions of OpenBLAS

Comment (omus) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch]: Re-reading your comments again I see I was supposed to remote the skylakex patch and apply the GCC patch. I'm unaware of how this is done with BB so I'll just give it a shot and let the CI sort out if it's correct.

Comment (giordano) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch]: I can have a look later today

Comment (omus) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch]: I've attempted to patch the avx512fintrin.h GCC header file. The patch applies cleanly on GCC 7.1.0 and 8.1.0 so I'm now attempting a CI run.

**Comment by giordano** [comment]
Now the license file isn't automatically installed anymore, so you need to do it manually: you need to run `install_license LICENSE` in the `${WORKSPACE}/srcdir/OpenBLAS*/` directory

File: O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch (hunk 1 above)
Comment (omus) [reply O/OpenBLAS/OpenBLAS32@0.3.13/bundled/patches/dasum_microk_skylakex-typecast.patch]: I changed OpenBLAS 0.3.12 to patch the GCC header too

**Comment by omus** [comment]
lol, too many jobs with the 0.3.12 changes. I'll drop that commit here

File: O/OpenBLAS/OpenBLASHighCoreCount@0.3.13/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,17 @@
+using BinaryBuilder
+
+include("../common.jl")
+
+# Collection of sources required to build OpenBLAS
+name = "OpenBLASHighCoreCount"
+version = v"0.3.13"
+
+sources = openblas_sources(version)
+script = openblas_script(num_64bit_threads=128)
```
Comment (ViralBShah) [inline O/OpenBLAS/OpenBLASHighCoreCount@0.3.13/build_tarballs.jl:10]: Should this number be bumped higher?

@chriselrod thoughts?

Comment (chriselrod) [reply O/OpenBLAS/OpenBLASHighCoreCount@0.3.13/build_tarballs.jl:10]: The Ampere Altra has 80 cores, and you can get [two sockets per motherboard](https://www.anandtech.com/show/16315/the-ampere-altra-review), but for x64 I think dual socket Epyc (for 128 cores) is the maximum.

What's the cost of increasing the maximum number of threads?

Comment (giordano) [reply O/OpenBLAS/OpenBLASHighCoreCount@0.3.13/build_tarballs.jl:10]: > What's the cost of increasing the maximum number of threads?

As far as I understand, bore memory allocated by OpenBLAS

**Comment by giordano** [comment]
Ok, I think we can go now?

**Comment by omus** [comment]
> Ok, I think we can go now?

We're good from my end. I can revise the patches for OpenBLAS 0.3.12 in another PR later to close: https://github.com/JuliaPackaging/Yggdrasil/issues/2389

**Comment by omus** [comment]
Should I update this to use symlinks for the patches?

**Pull request merged**